pkg/sfsturbo/nodeserver.go

```diff
@@ -17,142 +17,223 @@ limitations under the License.
-	}
-	if len(req.GetVolumeId()) == 0 {
-		return nil, status.Error(codes.InvalidArgument, "Volume ID missing in request")
-	}
-	if len(req.GetTargetPath()) == 0 {
-		return nil, status.Error(codes.InvalidArgument, "Target path missing in request")
-	}
-
-	target := req.GetTargetPath()
-	if len(target) == 0 {
-		return nil, status.Error(codes.InvalidArgument, "Target path not provided")
-	}
-
-	//Get Volume
+func (ns *nodeServer) NodePublishVolume(_ context.Context, req *csi.NodePublishVolumeRequest) (
+	*csi.NodePublishVolumeResponse, error) {
+	log.Infof("NodePublishVolume: called with args %v", protosanitizer.StripSecrets(*req))
+	capability := req.GetVolumeCapability()
 	volumeID := req.GetVolumeId()
-	client, err := ns.Driver.cloud.SFSTurboV1Client()
-    if err != nil {
-		klog.V(3).Infof("NodePublishVolume Failed to create SFS v2 client: %v", err)
-		return nil, status.Error(codes.Internal, err.Error())
-    }
-
-	// wait create new share
-	err = waitForShareStatus(client, volumeID)
-	if err != nil {
+	targetPath := req.GetTargetPath()
+	if err := nodePublishValidation(capability, volumeID, targetPath); err != nil {
 		return nil, err
 	}
 
-	share, err := getShare(client, volumeID)
+	cloud := ns.Driver.cloud
+	share, err := services.GetShare(cloud, volumeID)
 	if err != nil {
-		if _, ok := err.(golangsdk.ErrDefault404); ok {
-			return nil, status.Error(codes.NotFound, fmt.Sprintf("NodePublishVolume Volume %s not found", volumeID))
+		if common.IsNotFound(err) {
+			return nil, status.Errorf(codes.NotFound, "Share %s has already been deleted.", volumeID)
 		}
-		return nil, status.Error(codes.Internal, fmt.Sprintf("NodePublishVolume %v", err))
+		return nil, status.Errorf(codes.Internal, "Failed to query share: %s, error: %v", volumeID, err)
 	}
 
 	//Get Volume export location
-	source := share.ExportLocation
-	if source == "" {
-		return nil, status.Error(codes.Internal, fmt.Sprintf("NodePublishVolume Volume %s location not found", volumeID))
+	exportLocation := share.ExportLocation
+	if len(exportLocation) == 0 {
+		return nil, status.Errorf(codes.Internal, "Not found export location from volume %s", volumeID)
 	}
 
-	mountOptions := "noresvport,nolock"
-	if req.GetReadonly() {
-		mountOptions += ",ro"
+	if isMounted(targetPath) {
+		log.Infof("NodePublishVolume: %s has already mounted", targetPath)
+		return &csi.NodePublishVolumeResponse{}, nil
 	}
 
-	if isMounted(target) {
-		klog.V(2).Infof("NodePublishVolume: %s is already mounted", target)
-		return &csi.NodePublishVolumeResponse{}, nil
+	if err := makeDir(targetPath); err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to make dir: %s, error: %v", targetPath, err)
 	}
 
-	klog.V(2).Infof("NodePublishVolume: creating dir %s", target)
-	if err := makeDir(target); err != nil {
-		return nil, status.Errorf(codes.Internal, "Could not create dir %q: %v", target, err)
+	mountOptions := "noresvport,nolock"
+	if req.GetReadonly() {
```

Comment:
does SFS Turbo support `rw`

Reply:
By default, it is mounted in read-write mode

Reply:
done